x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java

```diff
@@ -507,12 +518,26 @@ public boolean isActive() {
     /**
      * Checks whether the given feature is allowed, tracking the last usage time.
      */
+    @SuppressForbidden(reason = "Argument to Math.abs() is definitely not Long.MIN_VALUE")
     public boolean checkFeature(Feature feature) {
         boolean allowed = isAllowed(feature);
         LongAccumulator maxEpochAccumulator = lastUsed.get(feature);
+        long now = System.currentTimeMillis();
+        long licenseExpirationDate = getLicenseExpiryDate();
         if (maxEpochAccumulator != null) {
             maxEpochAccumulator.accumulate(epochMillisProvider.getAsLong());
         }
+
+        if (feature.minimumOperationMode.compareTo(OperationMode.BASIC) > 0 &&
+            now >  licenseExpirationDate - LICENSE_EXPIRATION_WARNING_PERIOD.getMillis()) {
+            final long days = TimeUnit.MILLISECONDS.toDays(licenseExpirationDate - now);
+            final String expiryMessage = days == 0? "expires today":
+                (days > 0? String.format(Locale.ROOT, "will expire in [%d] days", days):
+                    String.format(Locale.ROOT, "has expired [%d] days ago", Math.abs(days)));
```

Comment:
I think we should be doing the first round of calculations in millis, not in days otherwise we potentially says "expires today" _after_ the expiry date (because it's less than 1 day expired)

```suggestion
    
        if (feature.minimumOperationMode.compareTo(OperationMode.BASIC) > 0) {
            final long expiryMillis = licenseExpirationDate - now;        
            if (expiryMillis < LICENSE_EXPIRATION_WARNING_PERIOD.getMillis()) {
                final String expiryMessage;
                if (expiryMillis <= 0) {
                    expiryMessage = "expired on [" + DATE_FORMATTER.formatMillis(expirationMillis) + "]";
                } else {
                    final long days = TimeUnit.MILLISECONDS.toDays(expiryMillis);
                    expiryMessage = days == 0 ? "expires today": 
                       String.format(Locale.ROOT, "will expire in [%d] days", days);
                }
```

Reply:
Done